File: algorithms/implementation/sock-merchant/sock-merchant-javascript/js/sockMerchant.js
// # O problema "Sock Merchant" no HackerRank 
// # é um desafio de programação que pede aos 
// # participantes para determinar quantos 
// # pares de meias idênticas podem ser formados 
// # a partir de um conjunto de meias. Aqui 
// # está uma descrição geral do problema:

// # Tarefa:
// # Dada uma lista de números inteiros 
// # representando as cores das meias, a tarefa 
// # é contar quantos pares de meias idênticas 
// # podem ser formados. Cada número inteiro 
// # representa uma cor diferente de meia.

// // Função que conta o número de pares de meias idênticas

// export function sockMerchant(n, ar) {

//     // Ordena a array em ordem crescente
//     ar.sort((a, b) => a - b);

//     // Inicializa variáveis para contar pares e armazenar a meia atual
//     let contador = 0;
//     let atual = 0;

//     // Itera sobre cada elemento na array
//     for(let elemento of ar) {

//         // Verifica se o elemento é igual à meia atual
//         if (elemento === atual) {
            
//             // Se sim, incrementa o contador de pares
//             contador += 1;

//             // Reinicia a variável atual para 0
//             atual = 0;

//             // Pula para a próxima iteração do loop
//             continue;

//         }

//         // Se o elemento não for igual à meia atual, atualiza a variável atual
//         atual = elemento;
//     }

//     // Retorna o número total de pares de meias idênticas
//     return contador;
    
// }


// // Função que conta o número de pares de meias idênticas

// export function sockMerchant(n, ar) {

//     // Usa o método reduce para contar as ocorrências de cada valor na array
//     const contador = ar.reduce((a, i) => {

//         // Se a chave já existe, incrementa o valor; caso contrário, inicializa com 1
//         a[i] = a[i] ? a[i] += 1 : 1;

//         // Retorna o objeto acumulado (contagem)
//         return a;

//     }, []);

//     // Usa o método reduce para calcular o número total de pares
//     const pares = contador.reduce((a, i) => {

//         // Adiciona a quantidade inteira da divisão por 2 ao acumulador
//         return a + Math.floor(i /2);

//     }, 0);

//     // Retorna o número total de pares de meias idênticas
//     return pares;
// }


// Função que conta o número de pares de meias idênticas

export function sockMerchant(n, ar) {

    // Calcula o índice da última meia na array
    let numMeia = n - 1;

    // Classifica a array em ordem crescente (altera o array original)
    let classificarArr = ar.sort((a, b) => a - b);

    // Inicializa a variável para contar o número total de pares
    let pares = 0;

    // Itera sobre a array para contar os pares de meias idênticas
    for (let index = 0; index < numMeia; index++) {
        
        // Verifica se a meia atual é idêntica à próxima meia
        if (classificarArr[index] === classificarArr[index + 1]) {
            
            // Se sim, incrementa o contador de pares
            pares += 1;

            // Avança o índice em mais uma posição para pular a próxima meia
            index += 1;
        }
        
    }

    // Retorna o número total de pares de meias idênticas
    return pares;
}

File: algorithms/implementation/sock-merchant/sock-merchant-python/tests/sockMerchant.py
# O problema "Sock Merchant" no HackerRank 
# é um desafio de programação que pede aos 
# participantes para determinar quantos 
# pares de meias idênticas podem ser formados 
# a partir de um conjunto de meias. Aqui 
# está uma descrição geral do problema:

# Tarefa:
# Dada uma lista de números inteiros 
# representando as cores das meias, a tarefa 
# é contar quantos pares de meias idênticas 
# podem ser formados. Cada número inteiro 
# representa uma cor diferente de meia.

# # Função que conta o número de pares de meias idênticas

# def sockMerchant(n, ar):
    
#     # Inicializa a variável para contar o número de pares
#     contar = 0
    
#     # Converte a lista de meias em um conjunto para obter cores únicas
#     conjunto = set(ar)
    
#     # Itera sobre cada cor única presente nas meias
#     for i in conjunto:
        
#         # Conta quantas vezes a cor aparece na lista original
#         elemento = ar.count(i)
        
#         # Verifica se há pelo menos dois elementos da mesma cor para formar um par
#         if elemento % 2 == 0 and elemento > 1:
            
#             # Adiciona a metade do número de elementos ao contador
#             contar += (elemento / 2)
            
#         # Verifica se há pelo menos dois elementos da mesma cor para formar um par
#         elif elemento % 2 != 0 and elemento > 1:
            
#             # Adiciona a parte inteira da divisão por dois ao contador
#             contar += (elemento // 2)
    
#     # Converte o contador para inteiro e retorna o resultado
#     return int(contar)

# Função que conta o número de pares de meias idênticas

# def sockMerchant(n, ar):
    
#     # Inicializa um dicionário para armazenar a contagem de cada tipo de meia
#     meias = {}
    
#     # Itera sobre cada meia na lista
#     for meia in ar:
        
#         # Verifica se a meia já está no dicionário
#         if meias.get(meia, None):
            
#             # Se sim, incrementa a contagem
#             meias[meia] += 1
        
#         else:
            
#             # Se não, adiciona a meia ao dicionário com contagem 1
#             meias[meia] = 1
            
#     # Inicializa a variável para contar o número total de pares
#     pares = 0
    
#     # Itera sobre as contagens de meias no dicionário
#     for conta in meias.values():
        
#         # Adiciona a quantidade inteira da divisão por 2 ao contador de pares
#         pares += conta // 2
    
#     # Retorna o número total de pares de meias idênticas
#     return pares


# Importa a classe Counter do módulo collections

from collections import Counter

# Função que conta o número de pares de meias idênticas

def sockMerchant(n, ar):
    
    # Inicializa uma variável para contar o número total de pares
    contador = 0
    
    # Cria um dicionário que conta a ocorrência de cada elemento na lista de meias
    dicionario = Counter(ar)
    
    # Itera sobre as chaves (elementos únicos) no dicionário
    for chave in dicionario:
        
        # Adiciona a quantidade inteira da divisão por 2 ao contador de pares
        contador += dicionario[chave] // 2
    
    # Retorna o número total de pares de meias idênticas
    return contador